### PR TITLE
Feature morph task

### PR DESCRIPTION
**Pull request by thematrixmaster**
Created new task for gfn training based on morphology cosine similarity reward

### CODE REVIEW

File: src/gflownet/models/mmc.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        super().__init__()
+        with hydra.initialize_config_dir(config_dir=cfg_dir, version_base=None):
+            cfg = hydra.compose(config_name=cfg_name)
+        model: LightningModule = utils.instantiate_model(cfg)
+        model = model.load_from_checkpoint(ckpt_path, map_location=worker)
+        model = model.to(worker)
+        self.model = model.eval()
+
+        # TODO make optional and load from arg path
+        self.latents = np.load("/home/mila/s/stephen.lu/gfn_gene/res/mmc/puma_embeddings.npz")
```
Comment (michalkoziarski) [inline src/gflownet/models/mmc.py]: So what exactly is contained in this file? Structure embeddings, joint embeddings, something else? Could you maybe add a comment for now (while it's hardcoded)?

File: src/gflownet/models/mmc.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+    def log_target_properties(self, target, target_latent, mode="joint"):
+        fig, ax = plt.subplots(figsize=(5, 5))
+
+        # compute cosine similarity between target modality and all struct latents
+        cosine_sim = cosine_similarity(target_latent, self.latents["struct"])
+        ax.hist(cosine_sim.flatten(), bins=50)
+        ax.set_title(f"Cosine similarity to target")
+
+        if mode == "joint":
```
Comment (michalkoziarski) [inline src/gflownet/models/mmc.py]: Why is this only being done in "joint" mode?

File: src/gflownet/models/mmc.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        self.model = model.eval()
+
+        # TODO make optional and load from arg path
+        self.latents = np.load("/home/mila/s/stephen.lu/gfn_gene/res/mmc/puma_embeddings.npz")
+
+    def log_target_properties(self, target, target_latent, mode="joint"):
+        fig, ax = plt.subplots(figsize=(5, 5))
+
+        # compute cosine similarity between target modality and all struct latents
+        cosine_sim = cosine_similarity(target_latent, self.latents["struct"])
```
Comment (michalkoziarski) [inline src/gflownet/models/mmc.py]: I am confused by this a bit (probably because I don't know what `self.latents` are - I assumed that they would be the latents of the targets in the dataset)?

File: src/gflownet/models/mmc.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+        self.wrap_for_mp = True
+
+        self.encoders = {k: wrap_model_fn(v) for (k, v) in self.model.encoders.items()}
+        self.projectors = {k: wrap_model_fn(v) for k, v in self.model.projectors.items()}
+        self.model.encoders = None
+        self.model.projectors = None
+
+        if self.model.__class__.__name__ == "GMC_PL":
+            self.encoder_joint = self.encoders["joint"]
```
Comment (michalkoziarski) [inline src/gflownet/models/mmc.py:70]: Type of encoder should probably be a parameter?

**Review by michalkoziarski — COMMENTED** [review]
I mostly added a few questions about things that confused me.

File: src/gflownet/models/mmc.py (hunk 1 above)
Comment (thematrixmaster) [reply src/gflownet/models/mmc.py]: This contains the struct embeddings of all molecules in the puma dataset.

File: src/gflownet/models/mmc.py (hunk 2 above)
Comment (thematrixmaster) [reply src/gflownet/models/mmc.py]: I initially figured that maybe at some point we won't have access to the target struct modality, but you're right that for now I should just plot whenever it's available.

File: src/gflownet/models/mmc.py (hunk 3 above)
Comment (thematrixmaster) [reply src/gflownet/models/mmc.py]: yeah sorry, I am plotting the cosine similarity between the target latent and the `struct` latents of all molecules in the training dataset to get a sense of how "difficult" it is to find similar representations.

File: src/gflownet/models/mmc.py (hunk 4 above)
Comment (thematrixmaster) [reply src/gflownet/models/mmc.py:70]: Yeah, this bit here was very nasty because I realized that I needed to superclass the GMC model class to pass all the submodules (encoders, projectors, etc.) of the proxy network through a message passing object to get the multiple workers to work. So what I'm doing here is essentially rewriting parts of the GMC_PL class methods that I couldn't easily call `super.method()` on. I'll clean it up as soon as I can!
